expression/expression.go

```diff
@@ -698,3 +709,155 @@ func wrapWithIsTrue(ctx sessionctx.Context, keepNull bool, arg Expression) (Expr
 	}
 	return FoldConstant(sf), nil
 }
+
+func getDatumBound(retType *types.FieldType, rType RoundingType) types.Datum {
+	if rType == Ceiling {
+		switch retType.Tp {
+		case mysql.TypeShort:
+			if mysql.HasUnsignedFlag(retType.Flag) {
+				return types.NewUintDatum(math.MaxUint16)
+			}
+			return types.NewIntDatum(math.MaxInt16)
+		case mysql.TypeLong:
+			if mysql.HasUnsignedFlag(retType.Flag) {
+				return types.NewUintDatum(math.MaxUint32)
+			}
+			return types.NewIntDatum(math.MaxInt32)
+		case mysql.TypeLonglong:
+			if mysql.HasUnsignedFlag(retType.Flag) {
+				return types.NewUintDatum(math.MaxUint64)
+			}
+			return types.NewIntDatum(math.MaxInt64)
+		case mysql.TypeFloat:
+			return types.NewFloat32Datum(math.MaxFloat32)
+		case mysql.TypeDouble:
+			return types.NewFloat64Datum(math.MaxFloat64)
+		case mysql.TypeNewDecimal:
+			return types.NewDecimalDatum(types.NewMaxOrMinDec(false, retType.Flen, retType.Decimal))
+		}
+	}
+	switch retType.Tp {
+	case mysql.TypeShort:
+		if mysql.HasUnsignedFlag(retType.Flag) {
+			return types.NewUintDatum(0)
+		}
+		return types.NewIntDatum(math.MinInt16)
+	case mysql.TypeLong:
+		if mysql.HasUnsignedFlag(retType.Flag) {
+			return types.NewUintDatum(0)
+		}
+		return types.NewIntDatum(math.MinInt32)
+	case mysql.TypeLonglong:
+		if mysql.HasUnsignedFlag(retType.Flag) {
+			return types.NewUintDatum(0)
+		}
+		return types.NewIntDatum(math.MinInt64)
+	case mysql.TypeFloat:
+		return types.NewFloat32Datum(-math.MaxFloat32)
+	case mysql.TypeDouble:
+		return types.NewFloat64Datum(-math.MaxFloat64)
+	case mysql.TypeNewDecimal:
+		return types.NewDecimalDatum(types.NewMaxOrMinDec(true, retType.Flen, retType.Decimal))
+	}
+	return types.Datum{}
+}
+
+// The function is for expression's reverse evaluation.
+// Here is an example for what's effort for the function: CastRealAsInt(t.a),
+// 		if the type of column `t.a` is mysql.TypeDouble, and there is a row that t.a == MaxFloat64
+// 		then the cast function will arrive a result MaxInt64. But when we do the reverse evaluation,
+//      if the result is MaxInt64, and the rounding type is ceiling. Then we should get the MaxFloat64
+//      instead of float64(MaxInt64).
+// Another example: cast(1.1 as signed) = 1,
```

Comment:
Could we really reverse it? Since there are multiply float values that could cast to 1.

Reply:
No, we can't really reverse it. All we need are just the range-bound for this situation, the example `cast(1.1 as signed) = 1` bounds are `2.0` for `Ceiling` and `1.0` for `Floor`.